arbnode/espresso_key_manager.go

```diff
@@ -0,0 +1,77 @@
+package arbnode
+
+import (
+	"crypto/ecdsa"
+	"crypto/rand"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/crypto"
+	"github.com/ethereum/go-ethereum/ethclient"
+	"github.com/ethereum/go-ethereum/log"
+)
+
+type EspressoKeyManagerInterface interface {
+	HasRegistered() bool
+	Registry(signFunc func([]byte) ([]byte, error)) error
+	GetCurrentKey() []byte
+	Sign(message []byte) ([]byte, error)
+}
+
+var _ EspressoKeyManagerInterface = &EspressoKeyManager{}
+
+type EspressoKeyManager struct {
+	address common.Address
+	pubKey  []byte
+	privKey *ecdsa.PrivateKey
+
+	hasRegistered bool
+	client        *ethclient.Client
+}
+
+func NewEspressoKeyManager(registryAddr common.Address, client *ethclient.Client) *EspressoKeyManager {
+	privKey, err := ecdsa.GenerateKey(crypto.S256(), rand.Reader)
+	if err != nil {
+		panic(err)
+	}
+
+	pubKey, ok := privKey.Public().(*ecdsa.PublicKey)
+	if !ok {
+		panic("failed to get public key")
+	}
+	pubKeyBytes := crypto.CompressPubkey(pubKey)
+
+	return &EspressoKeyManager{
+		address: registryAddr,
+		pubKey:  pubKeyBytes,
+		privKey: privKey,
+		client:  client,
+	}
+}
+
+func (k *EspressoKeyManager) HasRegistered() bool {
```

Comment:
`HasRegistered` and `Registry` should both call the contract and see if the key has been registered

Reply:
I think Jeremy's intention is that `Registry` will call this new contract to register the key after using the `signFunc` to get the attestation quote over the public key.

`HasRegistered` probably should track the contract too even if we are keeping some local in memory state about the status of the keys registration.